src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -374,85 +376,257 @@ internal static BigInteger ParseBigInteger(string value, NumberStyles style, Num
+            int totalDigitCount = 0;
+            int numberScale = number.scale;
+
+            const int MaxPartialDigits = 9;
+            const uint TenPowMaxPartial = 1000000000;
 
-            BigInteger ten = 10;
-            value = 0;
-            while (--i >= 0)
+            try
             {
-                value *= ten;
-                if (number.digits[cur] != '\0')
+                foreach (ReadOnlyMemory<char> digitsChunk in number.digits.GetChunks())
+                {
+                    bool isValid = ProcessChunk(digitsChunk.Span, ref currentBuffer);
+                    if (!isValid)
+                        return false;
+                }
+
+                if (partialDigitCount > 0)
+                    MultiplyAdd(ref currentBuffer, s_uint32PowersOfTen[partialDigitCount], partialValue);
+
+                int trailingZeroCount = numberScale - totalDigitCount;
+
+                while (trailingZeroCount >= MaxPartialDigits)
                 {
-                    value += number.digits[cur++] - '0';
+                    MultiplyAdd(ref currentBuffer, TenPowMaxPartial, 0);
+                    trailingZeroCount -= MaxPartialDigits;
                 }
+
+                if (trailingZeroCount > 0)
+                    MultiplyAdd(ref currentBuffer, s_uint32PowersOfTen[trailingZeroCount], 0);
+
+                if (currentBufferSize == 0)
+                {
+                    sign = 0;
+                    bits = null;
+                }
+                else if (currentBufferSize == 1 && currentBuffer[0] <= int.MaxValue)
+                {
+                    sign = (int)(number.sign ? -currentBuffer[0] : currentBuffer[0]);
+                    bits = null;
+                }
+                else
+                {
+                    sign = number.sign ? -1 : 1;
+                    bits = currentBuffer.Slice(0, currentBufferSize).ToArray();
+                }
+
+                return true;
             }
-            while (number.digits[cur] != '\0')
+            finally
             {
-                if (number.digits[cur++] != '0') return false; // Disallow non-zero trailing decimal places
+                if (arrayFromPool != null)
+                    ArrayPool<uint>.Shared.Return(arrayFromPool);
             }
-            if (number.sign)
+
+            bool ProcessChunk(ReadOnlySpan<char> chunkDigits, ref Span<uint> _currentBuffer)
             {
-                value = -value;
+                int remainingIntDigitCount = Math.Max(numberScale - totalDigitCount, 0);
+                ReadOnlySpan<char> intDigitsSpan = chunkDigits.Slice(0, Math.Min(remainingIntDigitCount, chunkDigits.Length));
+
+                for (int i = 0; i < intDigitsSpan.Length; i++)
+                {
+                    char digitChar = chunkDigits[i];
+                    if (digitChar == '\0')
+                        return true;
+
+                    partialValue = partialValue * 10 + (uint)(digitChar - '0');
```

Comment:
The only alternatives to this that I can think of are:
- Create our own struct to hold the state variables. This is the same as what the compiler does for captured variables, so there would be no performance difference.
- Pass them individually as ref parameters. This would probably be worse as the wrapper struct is being passed as a single ref.
Is there anything else that can be done?

Reply:
If it doesn't impact perf, it's fine.

If it does measurably impact perf, it's likely due to every access being via an indirection, and you can avoid that either by storing the relevant state into locals, operating on the locals, and then storing the locals back into the fields, or by passing in and returning the relevant state by value.

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunHiUGAWQAUASg5ceUgJYA7DA2kMAvAwAMAbhqSp3AJL4ADgBtpYaRgCCYMDFy4xW6rp4BlDNBh75EADIQwbGMHUScXbiVVAAU8XEtZABMAJRgMAFcoWQAhAE8ANSC0mGFpUJ1dcql+BgMTMwtrW3sxSp5OZ3CpADNoBhL5BlkVdQ1BhgAeBnJNQYBqWdFWl3bOzqVZ1Vkw1e4AXyWGfY6XA+r3T28Pf0DgluPdFZ3FAeMAoL1h6W2dnqg+uQUQ1UMyGk2mo1k80W906jyePFeNw+G0G31WR3hkQYiPeaKkGPCBwBDBi9niyVSGWy+UKxWJpQOcNWv3+AyBIzGYJBUIOUiZmIYKK2vJ4BJ2xAA7Io8aLykddkA=

Reply:
@jfd16 Can you confirm that this feedback was resolved? Were you able to verify that there's no perf impact?

Reply:
> 
> 
> @jfd16 Can you confirm that this feedback was resolved? Were you able to verify that there's no perf impact?

This has been resolved.